Source/Core/Core/IOS/Network/KD/VFF/VFFUtil.cpp

```diff
@@ -0,0 +1,297 @@
+
+static IOS::HLE::FS::FileHandle* s_vff;
+
+static DRESULT vff_read(BYTE pdrv, BYTE* buff, LBA_t sector, UINT count);
+static DRESULT vff_write(BYTE pdrv, const BYTE* buff, LBA_t sector, UINT count);
+static DRESULT vff_ioctl(BYTE pdrv, BYTE cmd, void* buff);
+static DRESULT read_vff_header(FATFS* fs);
+
+enum : u16
+{
+  SECTOR_SIZE = 512,
+  VF_LITTLE_ENDIAN = 0xFFFE,
+  VF_BIG_ENDIAN = 0xFEFF
+};
+
+static FRESULT vff_mount(FATFS* fs)
+{
+  fs->fs_type = 0;     // Clear the filesystem object
+  fs->pdrv = (BYTE)0;  // Volume hosting physical drive
+
+  DRESULT ret = read_vff_header(fs);
+  if (ret != RES_OK)
+    return FR_DISK_ERR;
+
+  return FR_OK;
+}
+
+DRESULT read_vff_header(FATFS* fs)
+{
+  struct Header header;
+  if (!s_vff->Read(&header, 1))
+  {
+    ERROR_LOG_FMT(IOS_WC24, "Failed to read VFF header.");
+    return RES_ERROR;
+  }
+
+  u16 cluster_size = 0;
+  u16 cluster_count = 0;
+
+  switch (Common::swap16(header.endianness))
+  {
+  case VF_BIG_ENDIAN:
+    cluster_size = Common::swap16(header.clusterSize) * 16;
+    cluster_count = Common::swap32(header.volumeSize) / cluster_size;
+    break;
+  case VF_LITTLE_ENDIAN:
+    // TODO: Actually implement.
+    // RiiConnect24's VFF are the only seen instance of Little Endian as it uses an SDK tool.
+    // Another option is to just delete these VFF's and let the current channel create a Big Endian
+    // one.
+    return RES_ERROR;
+  }
+
+  if (cluster_count < 4085)
+  {
+    fs->fs_type = FS_FAT12;
+
+    u32 table_size = std::floor((cluster_count + 1) / 2) * 3;
+    table_size = (table_size + cluster_size - 1) & ~(cluster_size - 1);
+
+    // Fsize is the full table size divided by 512 (Cluster size).
+    fs->fsize = table_size / 512;
+  }
+  else if (cluster_count < 65525)
+  {
+    fs->fs_type = FS_FAT16;
+
+    u32 table_size = cluster_count * 2;
+    table_size = (table_size + cluster_size - 1) & ~(cluster_size - 1);
+    fs->fsize = table_size / 512;
+  }
+  else
+  {
+    ERROR_LOG_FMT(IOS_WC24, "VFF not FAT12 or 16! Cluster size: {}", cluster_size);
+    return RES_ERROR;
+  }
+
+  fs->n_fats = 2;
+  fs->csize = 1;
+
+  // Root directory entry is 4096 bytes long, with each entry being 32 bytes. 4096 / 32 = 128
+  fs->n_rootdir = 128;
+
+  u32 sysect = 1 + (fs->fsize * 2) + fs->n_rootdir / (512 / 32);
+
+  // cluster_count is the total count whereas this is the actual amount of clusters we can use
+  u32 actual_cluster_count = cluster_count - sysect;
+
+  fs->n_fatent = actual_cluster_count + 2;
+  fs->volbase = 0;
+  fs->fatbase = 1;
+  fs->database = sysect;
+  // Root directory entry
+  fs->dirbase = fs->fatbase + fs->fsize * 2;
+
+  // Initialize cluster allocation information
+  fs->last_clst = fs->free_clst = 0xFFFFFFFF;
+  fs->fsi_flag = 0x80;
+
+  fs->id = 0;
+  fs->cdir = 0;
+
+  return RES_OK;
+}
+
+DRESULT vff_read(BYTE pdrv, BYTE* buff, LBA_t sector, UINT count)
+{
+  const u64 offset = static_cast<u64>(sector) * SECTOR_SIZE - 480;
```

Comment:
This should probably check if `sector == 0` and abort if so, as the resulting offset is nonsense in that case.